janMarkusMilan/assignments/reverseLinked/lc92/lc92.js

```diff
@@ -0,0 +1,184 @@
+  let currentNode = head;
+  let start = head; // Represents the last list node before the start of our reversed linked list section
+
+  // Iterate through until we reach p
+  while (currentPos < p) {
+    start = currentNode; // Will end up pointing to the node before the reversed linked list section, or p - 1
+    currentNode = currentNode.next; // Advance current node (will end up equal to value at node p)
+    currentPos++; // Increment position (will end up equal to p)
+  }
+
+  let newList = null; // Initialize a new list
+  let tail = currentNode; // The beginning node in the target section of the linked list will end up as the tail once reversed
+
+  // Now iterate through the reversal while the currentPos is in the boundaries of p and q
+  while (currentPos >= p && currentPos <= q) {
+    const nextNode = currentNode.next;
+    currentNode.next = newList;
+    newList = currentNode; // currentNode will represent the head of the list we have built so far
+    currentNode = nextNode; // At the end, currentNode will equal the value at q + 1, which represents the start of the rest of the linked list
+    currentPos++;
+  }
+
+  start.next = newList; // The newList is the head of the reversed section
+  tail.next = currentNode; // currentNode is pointing at q + 1
+
+  // Most of the time, we could return head, but it won't work if p is 1, meaning we reverse starting from the beginning, so we need to check
+  if (p > 1) {
+    return head;
+  } else {
+    return newList;
+  }
+};
+
+/*
+Sample run of my solution:
+
+Linked List: (1) -> (2) -> (3) -> (4) -> (5) -> null, p = 2, q = 4
+
+currentPos = 1
+currentNode = head = (1)
+start = head = (1)
+
+Loop (iterate until we reach p):
+  Iteration 0:    start = currentNode = (1)
+                  currentNode = currentNode.next = (2)
+                  currentPos++ -> currentPos = 2
+  Since currentPos === p, close loop
+
+newList = null
+tail = currentNode = (2)
+
+Loop:
+  Iteration 0:  nextNode = currentNode.next = (3)
+                currentNode.next = newList = (2) -> null
+                newList = currentNode = (2) -> null
+                currentNode = nextNode = (3)
+                currentPos++ -> currentPos = 3
+  Iteration 1:  nextNode = currentNode.next = (4)
+                currentNode.next = newList = (3) -> (2) -> null
+                newList = currentNode = (3) -> (2) -> null
+                currentNode = nextNode = (4)
+                currentPos++ -> currentPos = 4
+  Iteration 2:  nextNode = currentNode.next = (5)
+                currentNode.next = newList = (4) -> (3) -> (2) -> null
+                newList = currentNode = (4) -> (3) -> (2) -> null
+                currentNode = nextNode = (5)
+                currentPos++ -> currentPos = 5
+  Since currentPos > q, close loop
+
+start.next = newList = (1) -> (4) -> (3) -> (2) -> null
+tail.next = currentNode = (1) -> (4) -> (3) -> (2) -> (5) -> null
+
+return head = (1) -> (4) -> (3) -> (2) -> (5) -> null
+*/
```

Comment:
This is an amazing step to help illustrate what is happening without a whiteboard! Be mindful of the time it will take you to complete this. On an assesment or on-site you may not have the time!